### PR TITLE
[Model] Support MiniMax-M2

### LINKED ISSUE #909: Development Roadmap (2026 Q2)
> **Q2 roadmap is frozen.** On 2026-07-26 the unchecked items were re-audited against `main`. Completed-but-stale entries were updated below; work that remains incomplete has been moved to the [2026 Q3 roadmap #1460](https://github.com/sgl-project/sglang-jax/issues/1460). Please do not add new Q2 scope here.

Here is the development roadmap of SGL-JAX for 2026 Q2.

### Focus

- **Mainstream Model Support:** Expand model coverage on TPU.
- **Kernel Optimization:** Optimize kernels for TPU v6e/v7 hardware.
- **Architecture Refactoring:** Improve framework internals for better maintainability and extensibility.
- **Community Development and Standardization:** Establish clearer contribution workflows, documentation, and CI infrastructure.
- **Performance Optimization in Pathways:** Optimize throughput for DeepSeek-R1-Distill-Qwen-1.5B and Qwen3-32B in Pathways.

### More Model Coverage

- [x] Support DeepSeek V3 @Bob-Chen222 [#950](https://github.com/sgl-project/sglang-jax/pull/950)
- [x] Support MiMo-V2-Flash and MiMo-V2-Pro @cjx0709 [#978](https://github.com/sgl-project/sglang-jax/pull/978)
- [x] Support Ling-2.5-1T (Bailing MOE 2.5) @JamesBrianD
- [x] Support Qwen3.5 @MokusMokun
- [x] Support Kimi-Linear @aolemila
- [x] Grok-2 baseline optimization: sequence parallel and sharding fixes @Bob-Chen222 [#952](https://github.com/sgl-project/sglang-jax/pull/952)

Optional models:

- [x] Support MiniMax M2
- [x] Support GLM-5
- [x] Support Kimi-K2.5

### Kernel Implementation and Optimization

- [x] Integrate and optimize RPA-v3 from tpu-inference @Iamleos
  - [x] Support attention sink
  - [x] Support Sliding Window Attention
- [x] Optimize EPMOE kernel
- [x] Optimize FusedMoe kernel [#1014](https://github.com/sgl-project/sglang-jax/issues/1014)
- [x] Implement and optimize Linear Attention Kernel
  - [x] Gated Linear Attention
  - [x] Kimi Delta Attention (KDA)
  - [x] Gated Delta Net (GDN)
  - [x] Implement the corresponding Attention Backend
- [x] TPU v7 quantized-matmul support/tuning for per-channel and block-wise formats ([#1191](https://github.com/sgl-project/sglang-jax/pull/1191), [#1259](https://github.com/sgl-project/sglang-jax/pull/1259), [#1320](https://github.com/sgl-project/sglang-jax/pull/1320))
- [ ] AllGather+MatMul fused kernel @Bob-Chen222 — moved to [Q3 #1460](https://github.com/sgl-project/sglang-jax/issues/1460); PR [#898](https://github.com/sgl-project/sglang-jax/pull/898) is not in `main`.

### Architecture Refactoring

- [x] Refactor scheduler to support single-controller Data-Parallel scheduling
- [x] SWA memory-pool enhancement @cjx0709 [#921](https://github.com/sgl-project/sglang-jax/pull/921)
- [ ] Refactor TPU-CPU data transfer patterns — moved to [Q3 #1460](https://github.com/sgl-project/sglang-jax/issues/1460), tracked by [#772](https://github.com/sgl-project/sglang-jax/issues/772), [#774](https://github.com/sgl-project/sglang-jax/issues/774), and [#1452](https://github.com/sgl-project/sglang-jax/issues/1452)
- [x] Refactor JIT pre-compilation
- [ ] Complete Memory Cache System Refactoring [#992](https://github.com/sgl-project/sglang-jax/issues/992) — moved to [Q3 #1460](https://github.com/sgl-project/sglang-jax/issues/1460) and [HiCache #1280](https://github.com/sgl-project/sglang-jax/issues/1280)
- [x] Refactor speculative decoding pipeline to reduce CPU-TPU data-transfer overhead
- [ ] Upgrade JAX/libtpu [#1025](https://github.com/sgl-project/sglang-jax/issues/1025) — target refreshed and moved to [Q3 #1460](https://github.com/sgl-project/sglang-jax/issues/1460)
- [ ] Finish DP feature parity and compatibility [#972](https://github.com/sgl-project/sglang-jax/issues/972) — moved to [Q3 #1460](https://github.com/sgl-project/sglang-jax/issues/1460)

### Community Development and Standardization

- [ ] Complete the PR review mechanism — moved to [Q3 #1460](https://github.com/sgl-project/sglang-jax/issues/1460)
  - [ ] Build module-based PR review assignments
  - [ ] Establish a dual-reviewer policy for new contributors and large features
  - [x] Add AI-assisted PR review automation ([#1174](https://github.com/sgl-project/sglang-jax/pull/1174), [#1186](https://github.com/sgl-project/sglang-jax/pull/1186))
- [ ] Complete community AI Skills — moved to [Q3 #1460](https://github.com/sgl-project/sglang-jax/issues/1460)
  - [ ] Documentation-writing skill
  - [ ] Test/CI-development skill
  - [x] Benchmark and profiling skills ([#1443](https://github.com/sgl-project/sglang-jax/pull/1443), [#1455](https://github.com/sgl-project/sglang-jax/pull/1455))
  - [ ] Code-review skill
  - [ ] Kernel development/tuning skill
- [ ] Continue weekly community meetings — moved to [Q3 #1460](https://github.com/sgl-project/sglang-jax/issues/1460)

### Performance Optimization in Pathways

The remaining Pathways work is moved to [Q3 #1460](https://github.com/sgl-project/sglang-jax/issues/1460): profile and reduce host-device transfer overhead, validate overlap, and close the decode fast-path gap. Tracking issues: [#772](https://github.com/sgl-project/sglang-jax/issues/772), [#774](https://github.com/sgl-project/sglang-jax/issues/774), [#1452](https://github.com/sgl-project/sglang-jax/issues/1452).

### Multimodal Performance Optimization

- [ ] Complete online/offline multimodal APIs and Tunix visual-API integration — moved to [Q3 #1460](https://github.com/sgl-project/sglang-jax/issues/1460) / [VLM #1464](https://github.com/sgl-project/sglang-jax/issues/1464)
- [ ] Refactor multimodal execution toward the standard single-controller scheduling path — moved to [Q3 #1460](https://github.com/sgl-project/sglang-jax/issues/1460) / [VLM #1464](https://github.com/sgl-project/sglang-jax/issues/1464)
- [ ] Complete more models:
  - [ ] Gemma 3 text/vision ([#1069](https://github.com/sgl-project/sglang-jax/issues/1069), [#816](https://github.com/sgl-project/sglang-jax/issues/816)) — moved to Q3
  - [ ] Gemma 4 VLM [#1023](https://github.com/sgl-project/sglang-jax/issues/1023) — text decoder landed in [#1323](https://github.com/sgl-project/sglang-jax/pull/1323); VLM work moved to Q3
  - [x] Flux 1.dev [#902](https://github.com/sgl-project/sglang-jax/pull/902)
  - [ ] Qwen-Image — moved to Q3
  - [ ] Qwen3-VL ([#1300](https://github.com/sgl-project/sglang-jax/issues/1300), [#782](https://github.com/sgl-project/sglang-jax/pull/782)) — moved to Q3
- [ ] Complete multimodal parallelism and acceleration — moved to Q3
  - [ ] Ulysses-SP / Ring-SP / CFG Parallel / TP ([#721](https://github.com/sgl-project/sglang-jax/issues/721))
  - [ ] Integrate Cache-DiT to accelerate video generation

### CI / Release / Maintenance

- [x] Release the PyPI package, Docker image, and GitHub Release through the two-phase release pipeline [#1219](https://github.com/sgl-project/sglang-jax/pull/1219)

### Migration destination

All remaining work above is tracked in the [2026 Q3 roadmap #1460](https://github.com/sgl-project/sglang-jax/issues/1460). The Q2 roadmap stays open as a frozen historical record while Q3 is still marked Draft.

### PR DESCRIPTION
**Pull request by zorrofox**
## Summary

Adds `MiniMaxM2ForCausalLM` (M2/M2.5/M2.7, 230B total / 10B active). Architecture: standard GQA (48h/8kv, head_dim=128) + sigmoid-gated MoE (256 experts, topk=8, `e_score_correction_bias`) + per-layer QK RMSNorm + partial RoPE (`rotary_dim=64`). Official checkpoint is FP8 block-wise 128×128 (same format as DeepSeek-V3).

Closes #909 (roadmap item "Support minimax m2").

## Implementation notes

This is sglang-jax's first **GQA + FP8-static + `tp > num_kv_heads`** combination, which surfaces several edge cases — all handled inside `minimax_m2.py` (no changes to shared code):

| Issue | Fix |
|---|---|
| `QuantizedLinear.from_linear` derives placeholder shapes from `LinearBase.output_size` *before* load-time `kv_head_padding`, so `weight_q` gets padded to `tp` heads but `weight_scale` stays at the unpadded placeholder shape | Size `k_proj`/`v_proj`/`k_norm`/`RadixAttention` for `max(num_kv_heads, tp)` at construction |
| `_apply_kv_head_padding` only matches `"k_proj"/"v_proj"` hf_keys, so the 1D per-layer `k_norm` scale loads unpadded | Post-load per-head tile on host (load replicated; multi-host sharded arrays are non-addressable for `np.asarray`) |
| Per-shard `k/v_proj` out_dim = `head_dim` = 128 = one N-block for any `tp >= num_kv_heads`, tripping the narrow-N guard | `allow_narrow_n_blockwise=True` via `patch_model_config`; GSM8K 5/5 confirms no accuracy loss |
| `qk_norm_type="per_layer"` norms over all heads jointly (`num_heads*head_dim`), variance must reduce across the tensor-sharded axis | `MiniMaxM2QKNorm` with scale sharded `P("tensor")`; GSPMD inserts the cross-shard all-reduce automatically (verified by `test_tp_variance_allreduce`) |

**Topology limitation**: `num_attention_heads=48` is not divisible by `tp=32` or `tp=64`, and sglang-jax has no q-head padding (only kv). M2/M2.5/M2.7 therefore run on **v6e-4/8/16** only among standard topologies.

## Tests (`test_minimax_m2.py`, CPU, <5s)

- Registry registration
- `MiniMaxM2QKNorm` reference + TP variance all-reduce
- Weight-mapping coverage vs HF `model.safetensors.index.json` (FP8 + BF16 paths)
- Mini-config FP8 + `tp>num_kv_heads` shape dry-run (catches all the above placeholder mismatches)

## E2E (v6e-16, FP8 native)

```bash
python -m sgl_jax.launch_server --model-path MiniMaxAI/MiniMax-M2 \
  --trust-remote-code --dtype bfloat16 --device tpu \
  --tp-size 16 --dp-size 2 --ep-size 16 \
  --mem-fraction-static 0.85 --max-running-requests 512 \
  --chunked-prefill-size 2048 --context-length 8192 --page-size 64 \
  --attention-backend fa --disable-radix-cache \
  --precompile-bs-paddings 1 8 64 100 128 160 200 256 \
  --precompile-token-paddings 256 1024 2048 \
  --nnodes 4 --node-rank ${RANK} --dist-init-addr ${MASTER}:5000
```

| cc (1K/1K random) | output tok/s | TPOT(ms) | TTFT(ms) |
|---|---|---|---|
| 1 | 115.7 | 8.6 | 102 |
| 128 | 2886.6 | 39.4 | 5104 |
| 200 | 3181.8 | 55.1 | 7915 |
| 300 | 3248.5 | 83.3 | 9327 |
| **400** | **3767.3** | 94.1 | 12311 |

GSM8K 5/5 PASS for all of M2/M2.5/M2.7. `/v1/chat/completions` requires #1277.

## Test plan
- [x] `pytest python/sgl_jax/test/models/test_minimax_m2.py` 6/6 PASS
- [x] v6e-16 e2e: load + precompile + GSM8K 5/5 (M2, M2.5, M2.7)
- [x] bench_serving 1K/1K cc=1..400 (M2; M2.5 cc=400 reproduces within 3%)
- [ ] CI lint + e2e-test-4-tpu

### CODE REVIEW

**Comment by pengchengneo** [comment]
Testing only with GSM8K for verification accuracy might not be enough. 

We've encountered situations where GSM8K scores were high, but the accuracy on inference datasets didn't match. 

Could you test with AIME-24/25 and MMLU-PRO datasets as well?

@zorrofox

**Comment by zorrofox** [comment]
@pengchengneo Thanks — ran additional accuracy evals on v6e-16 with `test/srt/run_eval.py` (simple_eval framework, `temperature=0.6, top_p=0.95` per M2 `generation_config.json`, single-sample):

| eval | n | max_tokens | sglang-jax | M2 official | Δ |
|---|---|---|---|---|---|
| **AIME25** | 30 | 40000 | **80.0%** | 78% | **+2.0pp** |
| GPQA-Diamond | 198 | 24000 | 68.7% | 78% | −9.3pp |
| MMLU (standard) | 500 | 4096 | 82.8% | (Pro=82%) | n/a (different dataset) |

**AIME25 80% > official 78%** confirms forward-pass numerical correctness end-to-end (FP8 narrow-N k/v_proj, per-layer QK norm cross-shard variance, partial RoPE all OK).

GPQA −9.3pp is within single-sample variance (std=0.46 over 198 samples; official is presumably avg@N). MMLU 82.8% is the **standard** MMLU; official reports **MMLU-Pro** which is a different (harder) dataset — `run_eval.py` doesn't have an `mmlu_pro` or `aime24` option yet.

Key gotcha discovered while running this: M2 is a reasoning model and AIME thinking can reach **30-40K tokens** — first attempt with `max_tokens=12000` truncated 100% of samples mid-`<think>` and scored a misleading 23.3%. Eval needs `--context-length ≥ 49152` and `--max-tokens ≥ 40000` for math benchmarks (added to PR description launch command).

Happy to run a longer GPQA avg@5 or add `mmlu_pro`/`aime24` to `run_eval.py` if you'd like a closer apples-to-apples comparison.

File: python/sgl_jax/test/models/test_minimax_m2.py
Comment (pengchengneo) [inline python/sgl_jax/test/models/test_minimax_m2.py:1]: this test just for dev phase ? 

shall we remove it before merge ?

**Comment by pengchengneo** [comment]
LGTM
do you have any more comments  @JamesBrianD @cjx0709 @Rodrian7  ?

**Comment by cjx0709** [comment]
@pengchengneo @zorrofox Great work, I don't have any more comments

**Review by cjx0709 — DISMISSED** [review]

File: python/sgl_jax/test/models/test_minimax_m2.py
Comment (zorrofox) [reply python/sgl_jax/test/models/test_minimax_m2.py:1]: Trimmed in 042d5bcf — dropped the dev-phase `WeightMappings` coverage and `Registry` check.

Kept two regression tests (128 lines, <5s on CPU) that I'd argue are worth merging:
- `TestMiniMaxM2QKNorm.test_tp_variance_allreduce` — the only proof that `qk_norm_type="per_layer"` (variance over the tensor-sharded `num_heads*head_dim` axis) is numerically correct under TP; this is M2-specific behavior not exercised by any other model.
- `TestMiniMaxM2AttentionShapes` — mini-config dry-run of the FP8-static + `tp>num_kv_heads` placeholder shapes. This is sglang-jax's first model with that combination; without this test a regression in `from_linear`/`kv_head_padding` would only surface after a 10-min 230GB load on TPU (which is exactly what happened during bring-up).

Happy to drop them too if you prefer.

Comment (pengchengneo) [reply python/sgl_jax/test/models/test_minimax_m2.py:1]: make sense, shall we add them to CI ? 
@zhengkezhou1 @Rodrian7 @JamesBrianD

**Review by pengchengneo — APPROVED** [review]
